design/1.0.md

```diff
@@ -0,0 +1,84 @@
+I recently opened a Kubernetes Enhancement Proposal to introduce `go-flow-levee` during testing as a defense against accidental credential logging.
+See [KEP-1933](https://github.com/kubernetes/enhancements/pull/1936) details.
+Such integration demands a stable API and well-defined scope.
+
+I propose the following.
+Portions may already be implemented, but are included here as a "big picture."
+
+# 1.0 Targets
+
+## Basic taint propagation detection
+
+Given a spec defining sources and sinks (below), diagnostics are reported for
+any direct call to a sink that includes a source.  Additionally, we detect any
+taint propagation from a source to a sink within the scope of a single function.
+
+## Analysis Configuration
+
+Configuration will consist of the following key components:
+
+* Identification of sources, sinks, and sanitizers
+* Identification of analysis scope, e.g. to skip third-party or testing packages.
+* Specify diagnostics to be ignored, e.g false-positives
+
+### Identification of sources, sinks, and sanitizers
+
+Much of our work is directly with `ssa.Value`s and `ssa.Instruction`s.
+Whether those values or instructions represent one element or another,
+ the configuration to identify these members should be as uniform as possible.
+In that light, specification should follow:
+
+#### ValueSpecifier
+A ValueSpecifier marks an `ssa.Value`, e.g., for identification as a source or as the safe result of a sanitizer.
+Values should be specifiable via any or all of:
+
+* Type / Field: Package path and type name, optional field name
+* Field tags
+* Scope: local variable, free variable, or global
+* Context: within a specific package or function call.  See CallSpecifier below.
+* Is a reference
+* Is "reference-like," e.g., slices, maps, chans.
+
```

Comment:
In addition to those, users might want to be able to specify specific _values_. For a 1.0, we could choose to support only basic literals (strings, ints, etc.), e.g. : 

`os.Getenv("PASSWORD")`.

Reply:
I had had "Const" in my noted but couldn't remember what I meant by it.  Good job.

It does occur to me (and this is perhaps implementation detail beyond the current scope) that if we're concerned about `os.Getenv("PASSWORD")`, we need to be also concerned about `func(s string) { os.Genenv(s) }`... I'm not sure how I'd want to specify "This is a source when called on this `const` term, or on any ambiguous term, but not when called on a `const` term like `"PATH"`."

Reply:
I think it would indeed be difficult to say "This is a source when called on an ambiguous term". I think the `func(s string)` case you mention would require cross-function analysis to do well. If the variable is defined within the same function we might be able to do better though, e.g. I think we could handle `func() { pass := "PASSWORD"; os.Getenv(pass) }`.